// @langchain/langgraph-checkpoint-mongodb/dist/checkpoint.cjs
let _langchain_langgraph_checkpoint = require("@langchain/langgraph-checkpoint");
//#region src/checkpoint.ts
function getStringConfigValue(name, value, { required = false } = {}) {
	if (value === void 0) {
		if (required) throw new Error(`Invalid configurable.${name}: expected a string`);
		return;
	}
	if (value === null || typeof value !== "string") throw new Error(`Invalid configurable.${name}: expected a string`);
	return value;
}
/**
* A LangGraph checkpoint saver backed by a MongoDB database.
*/
var MongoDBSaver = class extends _langchain_langgraph_checkpoint.BaseCheckpointSaver {
	client;
	db;
	checkpointCollectionName = "checkpoints";
	checkpointWritesCollectionName = "checkpoint_writes";
	enableTimestamps;
	get timestampOp() {
		return this.enableTimestamps ? { $currentDate: { upserted_at: true } } : {};
	}
	constructor({ client, dbName, checkpointCollectionName, checkpointWritesCollectionName, enableTimestamps }, serde) {
		super(serde);
		this.client = client;
		this.client.appendMetadata({ name: "langgraphjs_checkpoint_saver" });
		this.db = this.client.db(dbName);
		this.checkpointCollectionName = checkpointCollectionName ?? this.checkpointCollectionName;
		this.checkpointWritesCollectionName = checkpointWritesCollectionName ?? this.checkpointWritesCollectionName;
		this.enableTimestamps = enableTimestamps ?? false;
	}
	/**
	* Retrieves a checkpoint from the MongoDB database based on the
	* provided config. If the config contains a "checkpoint_id" key, the checkpoint with
	* the matching thread ID and checkpoint ID is retrieved. Otherwise, the latest checkpoint
	* for the given thread ID is retrieved.
	*/
	async getTuple(config) {
		const thread_id = getStringConfigValue("thread_id", config.configurable?.thread_id);
		if (thread_id === void 0) return;
		const checkpoint_ns = getStringConfigValue("checkpoint_ns", config.configurable?.checkpoint_ns) ?? "";
		const checkpoint_id = getStringConfigValue("checkpoint_id", config.configurable?.checkpoint_id);
		let query;
		if (checkpoint_id) query = {
			thread_id,
			checkpoint_ns,
			checkpoint_id
		};
		else query = {
			thread_id,
			checkpoint_ns
		};
		const result = await this.db.collection(this.checkpointCollectionName).find(query).sort("checkpoint_id", -1).limit(1).toArray();
		if (result.length === 0) return;
		const doc = result[0];
		const configurableValues = {
			thread_id,
			checkpoint_ns,
			checkpoint_id: doc.checkpoint_id
		};
		const checkpoint = await this.serde.loadsTyped(doc.type, doc.checkpoint.value("utf8"));
		const serializedWrites = await this.db.collection(this.checkpointWritesCollectionName).find(configurableValues).toArray();
		const pendingWrites = await Promise.all(serializedWrites.map(async (serializedWrite) => {
			return [
				serializedWrite.task_id,
				serializedWrite.channel,
				await this.serde.loadsTyped(serializedWrite.type, serializedWrite.value.value("utf8"))
			];
		}));
		return {
			config: { configurable: configurableValues },
			checkpoint,
			pendingWrites,
			metadata: await this.serde.loadsTyped(doc.type, doc.metadata.value("utf8")),
			parentConfig: doc.parent_checkpoint_id != null ? { configurable: {
				thread_id,
				checkpoint_ns,
				checkpoint_id: doc.parent_checkpoint_id
			} } : void 0
		};
	}
	/**
	* Retrieve a list of checkpoint tuples from the MongoDB database based
	* on the provided config. The checkpoints are ordered by checkpoint ID
	* in descending order (newest first).
	*/
	async *list(config, options) {
		const { limit, before, filter } = options ?? {};
		const query = {};
		const thread_id = getStringConfigValue("thread_id", config.configurable?.thread_id);
		const checkpoint_ns = getStringConfigValue("checkpoint_ns", config.configurable?.checkpoint_ns);
		if (thread_id) query.thread_id = thread_id;
		if (checkpoint_ns !== void 0) query.checkpoint_ns = checkpoint_ns;
		if (filter) Object.entries(filter).forEach(([key, value]) => {
			if (value !== null && typeof value === "object") throw new Error(`Invalid filter value for key "${key}": filter values must be primitives (string, number, boolean, or null)`);
			query[`metadata.${key}`] = value;
		});
		if (before?.configurable?.checkpoint_id !== void 0) query.checkpoint_id = { $lt: getStringConfigValue("checkpoint_id", before.configurable?.checkpoint_id) };
		let result = this.db.collection(this.checkpointCollectionName).find(query).sort("checkpoint_id", -1);
		if (limit !== void 0) result = result.limit(limit);
		for await (const doc of result) {
			const checkpoint = await this.serde.loadsTyped(doc.type, doc.checkpoint.value("utf8"));
			const metadata = await this.serde.loadsTyped(doc.type, doc.metadata.value("utf8"));
			yield {
				config: { configurable: {
					thread_id: doc.thread_id,
					checkpoint_ns: doc.checkpoint_ns,
					checkpoint_id: doc.checkpoint_id
				} },
				checkpoint,
				metadata,
				parentConfig: doc.parent_checkpoint_id ? { configurable: {
					thread_id: doc.thread_id,
					checkpoint_ns: doc.checkpoint_ns,
					checkpoint_id: doc.parent_checkpoint_id
				} } : void 0
			};
		}
	}
	/**
	* Saves a checkpoint to the MongoDB database. The checkpoint is associated
	* with the provided config and its parent config (if any).
	*/
	async put(config, checkpoint, metadata) {
		const thread_id = getStringConfigValue("thread_id", config.configurable?.thread_id, { required: true });
		const checkpoint_ns = getStringConfigValue("checkpoint_ns", config.configurable?.checkpoint_ns) ?? "";
		const parent_checkpoint_id = getStringConfigValue("checkpoint_id", config.configurable?.checkpoint_id);
		const checkpoint_id = checkpoint.id;
		const [[checkpointType, serializedCheckpoint], [metadataType, serializedMetadata]] = await Promise.all([this.serde.dumpsTyped(checkpoint), this.serde.dumpsTyped(metadata)]);
		if (checkpointType !== metadataType) throw new Error("Mismatched checkpoint and metadata types.");
		const doc = {
			parent_checkpoint_id,
			type: checkpointType,
			checkpoint: serializedCheckpoint,
			metadata: serializedMetadata
		};
		const upsertQuery = {
			thread_id,
			checkpoint_ns,
			checkpoint_id
		};
		await this.db.collection(this.checkpointCollectionName).updateOne(upsertQuery, {
			$set: doc,
			...this.timestampOp
		}, { upsert: true });
		return { configurable: {
			thread_id,
			checkpoint_ns,
			checkpoint_id
		} };
	}
	/**
	* Saves intermediate writes associated with a checkpoint to the MongoDB database.
	*/
	async putWrites(config, writes, taskId) {
		const thread_id = getStringConfigValue("thread_id", config.configurable?.thread_id, { required: true });
		const checkpoint_ns = getStringConfigValue("checkpoint_ns", config.configurable?.checkpoint_ns, { required: true });
		const checkpoint_id = getStringConfigValue("checkpoint_id", config.configurable?.checkpoint_id, { required: true });
		const operations = await Promise.all(writes.map(async ([channel, value], idx) => {
			const upsertQuery = {
				thread_id,
				checkpoint_ns,
				checkpoint_id,
				task_id: taskId,
				idx
			};
			const [type, serializedValue] = await this.serde.dumpsTyped(value);
			return { updateOne: {
				filter: upsertQuery,
				update: {
					$set: {
						channel,
						type,
						value: serializedValue
					},
					...this.timestampOp
				},
				upsert: true
			} };
		}));
		await this.db.collection(this.checkpointWritesCollectionName).bulkWrite(operations);
	}
	async deleteThread(threadId) {
		if (typeof threadId !== "string") throw new Error("Invalid threadId: expected a string");
		await this.db.collection(this.checkpointCollectionName).deleteMany({ thread_id: threadId });
		await this.db.collection(this.checkpointWritesCollectionName).deleteMany({ thread_id: threadId });
	}
};
//#endregion
exports.MongoDBSaver = MongoDBSaver;
