// @langchain/langgraph-checkpoint-mongodb/dist/store.js
import { MongoClient } from "mongodb";
import { BaseStore, InvalidNamespaceError } from "@langchain/langgraph-checkpoint";
//#region src/store.ts
/**
* Computes the namespace path prefixes for denormalization.
* E.g., ["a", "b", "c"] → ["a", "a/b", "a/b/c"]
*
* Stored on each document as `namespacePath` to enable prefix filtering
* in $vectorSearch, which does not support $expr or $slice.
* MongoDB array equality matches if any element equals the filter value,
* so filtering by "a/b" matches any document whose namespace starts with ["a", "b"].
*/
function computeNamespacePath(namespace) {
	const paths = [];
	for (let i = 1; i <= namespace.length; i++) paths.push(namespace.slice(0, i).join("/"));
	return paths;
}
/**
* Validates the provided namespace.
* @param namespace The namespace to validate.
* @throws {InvalidNamespaceError} If the namespace is invalid.
*/
function validateNamespace(namespace) {
	if (namespace.length === 0) throw new InvalidNamespaceError("Namespace cannot be empty.");
	for (const label of namespace) {
		if (typeof label !== "string") throw new InvalidNamespaceError(`Invalid namespace label '${label}' found in ${namespace}. Namespace labels must be strings, but got ${typeof label}.`);
		if (label.includes(".")) throw new InvalidNamespaceError(`Invalid namespace label '${label}' found in ${namespace}. Namespace labels cannot contain periods ('.').`);
		if (label === "") throw new InvalidNamespaceError(`Namespace labels cannot be empty strings. Got ${label} in ${namespace}`);
	}
	if (namespace[0] === "langgraph") throw new InvalidNamespaceError(`Root label for namespace cannot be "langgraph". Got: ${namespace}`);
}
/**
* A long-term persistent key-value store backed by MongoDB.
* Supports hierarchical namespace organization with Put, Get, ListNamespaces, and Search operations.
*/
var MongoDBStore = class MongoDBStore extends BaseStore {
	client;
	db;
	collectionName;
	enableTimestamps;
	ttl;
	embeddings;
	indexConfig;
	ownsClient = false;
	get timestampOp() {
		return this.enableTimestamps ? { $currentDate: { upserted_at: true } } : {};
	}
	constructor({ client, dbName, collectionName = "store", enableTimestamps, ttl, embeddings, indexConfig }) {
		super();
		this.client = client;
		this.client.appendMetadata({ name: "langgraphjs_store" });
		this.db = this.client.db(dbName);
		this.collectionName = collectionName;
		this.enableTimestamps = enableTimestamps ?? false;
		this.ttl = ttl;
		this.embeddings = embeddings;
		this.indexConfig = indexConfig;
	}
	/**
	* Factory method to create a MongoDBStore from a connection string.
	* Automatically creates and connects a MongoDB client, and calls
	* {@link start} to ensure required indexes exist.
	* @param connString MongoDB connection string
	* @param params Store configuration
	*/
	static async fromConnString(connString, params) {
		const client = new MongoClient(connString);
		await client.connect();
		const store = new MongoDBStore({
			...params,
			client
		});
		store.ownsClient = true;
		await store.start();
		return store;
	}
	/**
	* Execute a batch of operations (Put, Get, ListNamespaces, Search).
	* Operations are processed in order, but multiple consecutive Puts are batched.
	* PutOperations are deduplicated by (namespace, key) — last write wins.
	*/
	async batch(operations) {
		const results = new Array(operations.length);
		let i = 0;
		while (i < operations.length) {
			const op = operations[i];
			if ("value" in op) {
				const putIndices = [i];
				let j = i + 1;
				while (j < operations.length && "value" in operations[j]) {
					putIndices.push(j);
					j++;
				}
				await this.batchPuts(putIndices.map((idx) => ({
					index: idx,
					op: operations[idx]
				})), results);
				i = j;
			} else {
				const readOp = op;
				if ("key" in readOp && !("namespacePrefix" in readOp)) results[i] = await this.getOp(readOp);
				else if ("namespacePrefix" in readOp) results[i] = await this.searchOp(readOp);
				else results[i] = await this.listNamespacesOp(readOp);
				i++;
			}
		}
		return results;
	}
	/**
	* Execute a batch of Put operations with deduplication.
	*/
	async batchPuts(putOpsWithIndex, results) {
		const deduped = /* @__PURE__ */ new Map();
		for (const { index, op } of putOpsWithIndex) {
			const key = JSON.stringify({
				namespace: op.namespace,
				key: op.key
			});
			deduped.set(key, {
				index,
				op
			});
		}
		if (deduped.size > 0) {
			const bulkOps = [];
			const opsList = Array.from(deduped.values());
			const embeddingByOpIndex = /* @__PURE__ */ new Map();
			if (this.indexConfig && this.embeddings) {
				const textsToEmbed = [];
				for (let i = 0; i < opsList.length; i++) {
					const { op } = opsList[i];
					if (op.value === null || op.index === false) continue;
					let text;
					if (Array.isArray(op.index)) {
						const fields = {};
						for (const field of op.index) fields[field] = op.value?.[field];
						text = JSON.stringify(fields);
					} else if (this.indexConfig.embeddingKey) text = JSON.stringify(op.value?.[this.indexConfig.embeddingKey]);
					else text = JSON.stringify(op.value);
					textsToEmbed.push({
						opIndex: i,
						text
					});
				}
				if (textsToEmbed.length > 0) {
					const vectors = await this.embeddings.embedDocuments(textsToEmbed.map((t) => t.text));
					for (let j = 0; j < textsToEmbed.length; j++) embeddingByOpIndex.set(textsToEmbed[j].opIndex, vectors[j]);
				}
			}
			for (let i = 0; i < opsList.length; i++) {
				const { op } = opsList[i];
				const { namespace, key, value } = op;
				validateNamespace(namespace);
				if (value === null) {
					bulkOps.push({ deleteOne: { filter: {
						namespace,
						key
					} } });
					continue;
				}
				const now = /* @__PURE__ */ new Date();
				const doc = {
					namespace,
					namespaceStr: namespace.join("/"),
					key,
					value,
					updatedAt: now
				};
				if (this.indexConfig) doc.namespacePath = computeNamespacePath(namespace);
				if (this.ttl) doc.expiresAt = new Date(now.getTime() + this.ttl.defaultTtl * 1e3);
				const embeddingVector = embeddingByOpIndex.get(i);
				if (embeddingVector !== void 0) {
					const embeddingField = this.indexConfig.path ?? "embedding";
					doc[embeddingField] = embeddingVector;
				}
				bulkOps.push({ updateOne: {
					filter: {
						namespace,
						key
					},
					update: {
						$set: doc,
						$setOnInsert: { createdAt: now },
						...this.timestampOp
					},
					upsert: true
				} });
			}
			if (bulkOps.length > 0) await this.db.collection(this.collectionName).bulkWrite(bulkOps);
			for (const { index } of putOpsWithIndex) results[index] = void 0;
		}
	}
	/**
	* Retrieve an item by namespace and key.
	* If refreshOnRead is enabled, resets the TTL timer.
	*/
	async getOp(op) {
		const { namespace, key } = op;
		let doc;
		if (this.ttl?.refreshOnRead) {
			const now = /* @__PURE__ */ new Date();
			const updateDoc = { updatedAt: now };
			if (this.ttl) updateDoc.expiresAt = new Date(now.getTime() + this.ttl.defaultTtl * 1e3);
			doc = await this.db.collection(this.collectionName).findOneAndUpdate({
				namespace,
				key
			}, { $set: updateDoc }, { returnDocument: "after" });
		} else doc = await this.db.collection(this.collectionName).findOne({
			namespace,
			key
		});
		if (!doc) return null;
		return {
			value: doc.value,
			key: doc.key,
			namespace: doc.namespace,
			createdAt: doc.createdAt,
			updatedAt: doc.updatedAt
		};
	}
	/**
	* List namespaces in the store, filtered by matchConditions (prefix/suffix with wildcards).
	*
	* Called by BaseStore.listNamespaces() which converts user-facing prefix/suffix
	* options into MatchCondition objects. Each MatchCondition specifies:
	*   - matchType: "prefix" (match from start) or "suffix" (match from end)
	*   - path: array of labels, where "*" is a wildcard matching any value
	*
	* All conditions must be satisfied (AND logic).
	*
	* Uses $expr with $arrayElemAt for positional element matching because
	* wildcards and suffix matching require skipping positions or indexing
	* from the end, which dot-notation cannot express.
	*/
	async listNamespacesOp(op) {
		const { matchConditions, maxDepth, limit, offset } = op;
		const pipeline = [];
		if (matchConditions && matchConditions.length > 0) {
			const conditions = [];
			for (const condition of matchConditions) {
				const { matchType, path } = condition;
				if (matchType === "prefix") {
					const elemConditions = [];
					for (let i = 0; i < path.length; i++) if (path[i] !== "*") elemConditions.push({ $eq: [{ $arrayElemAt: ["$namespace", i] }, path[i]] });
					elemConditions.push({ $gte: [{ $size: "$namespace" }, path.length] });
					conditions.push({ $and: elemConditions });
				} else if (matchType === "suffix") {
					const elemConditions = [];
					for (let i = 0; i < path.length; i++) if (path[i] !== "*") elemConditions.push({ $eq: [{ $arrayElemAt: ["$namespace", -(path.length - i)] }, path[i]] });
					elemConditions.push({ $gte: [{ $size: "$namespace" }, path.length] });
					conditions.push({ $and: elemConditions });
				}
			}
			if (conditions.length > 0) pipeline.push({ $match: { $expr: conditions.length === 1 ? conditions[0] : { $and: conditions } } });
		}
		pipeline.push({ $group: { _id: "$namespace" } });
		if (maxDepth !== void 0) pipeline.push({ $match: { $expr: { $lte: [{ $size: "$_id" }, maxDepth] } } });
		pipeline.push({ $sort: { _id: 1 } });
		if (offset !== void 0 && offset > 0) pipeline.push({ $skip: offset });
		pipeline.push({ $limit: limit ?? 100 });
		return (await this.db.collection(this.collectionName).aggregate(pipeline).toArray()).map((doc) => doc._id);
	}
	/**
	* Search for items by namespace prefix and filter criteria.
	* Supports field-based filtering (structured search) on value fields.
	*/
	async searchOp(op) {
		const { namespacePrefix, filter, limit = 100, offset = 0, query } = op;
		if (query) {
			if (!this.indexConfig) throw new Error("Vector search (query parameter) requires indexConfig to be configured.");
			return this.vectorSearch(query, namespacePrefix, limit, offset);
		}
		const mongoQuery = {};
		if (namespacePrefix && namespacePrefix.length > 0) for (let idx = 0; idx < namespacePrefix.length; idx++) mongoQuery[`namespace.${idx}`] = namespacePrefix[idx];
		if (filter && Object.keys(filter).length > 0) for (const [field, condition] of Object.entries(filter)) if (typeof condition === "object" && condition !== null && !Array.isArray(condition)) for (const [operator, operatorValue] of Object.entries(condition)) switch (operator) {
			case "$eq":
				mongoQuery[`value.${field}`] = operatorValue;
				break;
			case "$ne":
				mongoQuery[`value.${field}`] = { $ne: operatorValue };
				break;
			case "$gt":
				mongoQuery[`value.${field}`] = { $gt: operatorValue };
				break;
			case "$gte":
				mongoQuery[`value.${field}`] = { $gte: operatorValue };
				break;
			case "$lt":
				mongoQuery[`value.${field}`] = { $lt: operatorValue };
				break;
			case "$lte":
				mongoQuery[`value.${field}`] = { $lte: operatorValue };
				break;
		}
		else mongoQuery[`value.${field}`] = condition;
		const docs = await this.db.collection(this.collectionName).find(mongoQuery).skip(offset).limit(limit).toArray();
		const items = [];
		for (const doc of docs) items.push({
			value: doc.value,
			key: doc.key,
			namespace: doc.namespace,
			createdAt: doc.createdAt,
			updatedAt: doc.updatedAt
		});
		return items;
	}
	/**
	* Perform vector similarity search using MongoDB $vectorSearch.
	*
	* Manual mode (embeddings configured): computes query vector via embedQuery(),
	* sends as queryVector to $vectorSearch.
	*
	* Auto mode (no embeddings): sends the query text directly via query.text,
	* MongoDB generates the query embedding server-side.
	*/
	async vectorSearch(query, namespacePrefix, limit, offset) {
		const path = this.indexConfig.path ?? "embedding";
		const vectorSearchStage = { $vectorSearch: {
			index: this.indexConfig.name,
			path,
			numCandidates: Math.min((limit + offset) * 20, 1e4),
			limit: limit + offset
		} };
		if (this.embeddings) vectorSearchStage.$vectorSearch.queryVector = await this.embeddings.embedQuery(query);
		else vectorSearchStage.$vectorSearch.query = { text: query };
		if (namespacePrefix.length > 0) vectorSearchStage.$vectorSearch.filter = { namespacePath: namespacePrefix.join("/") };
		const pipeline = [vectorSearchStage, { $addFields: { score: { $meta: "vectorSearchScore" } } }];
		if (this.embeddings) pipeline.push({ $project: { [path]: 0 } });
		if (offset > 0) pipeline.push({ $skip: offset });
		pipeline.push({ $limit: limit });
		return (await this.db.collection(this.collectionName).aggregate(pipeline).toArray()).map((doc) => ({
			value: doc.value,
			key: doc.key,
			namespace: doc.namespace,
			createdAt: doc.createdAt,
			updatedAt: doc.updatedAt,
			score: doc.score
		}));
	}
	/**
	* Initialize the store: creates the {namespace, key} unique index and an optional
	* TTL index.
	*/
	async start() {
		const collection = this.db.collection(this.collectionName);
		await collection.createIndex({
			namespaceStr: 1,
			key: 1
		}, { unique: true });
		if (this.ttl) await collection.createIndex({ expiresAt: 1 }, { expireAfterSeconds: 0 });
		if (this.indexConfig) {
			const fields = [];
			const path = this.indexConfig.path ?? "embedding";
			if (this.embeddings) fields.push({
				type: "vector",
				path,
				numDimensions: this.indexConfig.dims,
				similarity: this.indexConfig.similarityFunction ?? "cosine"
			});
			else fields.push({
				type: "autoEmbed",
				path,
				model: this.indexConfig.model,
				modality: this.indexConfig.modality ?? "text"
			});
			fields.push({
				type: "filter",
				path: "namespacePath"
			});
			for (const filterField of this.indexConfig.filters ?? []) fields.push({
				type: "filter",
				path: filterField
			});
			try {
				await collection.createSearchIndex({
					name: this.indexConfig.name,
					type: "vectorSearch",
					definition: { fields }
				});
			} catch (err) {
				if (!err?.message?.toLowerCase().includes("already exists")) throw err;
			}
		}
	}
	/**
	* Clean up the store. Closes the underlying MongoClient only if the store
	* created it (i.e. via {@link fromConnString}). When the client was supplied
	* to the constructor by the caller, the caller owns its lifecycle.
	*/
	async stop() {
		if (this.ownsClient) await this.client.close();
	}
};
//#endregion
export { MongoDBStore };
